### PR TITLE
NGSIv2 spec: typePattern

### PR DESCRIPTION
**Pull request by fgalan**
This PR includes the changes described in #1853 in the WIP .apib file of the NGSIv2 specification.

@kzangeli
@jmcanterafonseca

Feedback is welcome!

Merging deadline: Friday July 15th EoB.

### CODE REVIEW

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 1]
```diff
@@ -1595,11 +1599,11 @@ A subscription is represented by a JSON object with the following fields:
...
 A `subject` contains the following subfields:
 
-+ `entities`: A list of objects, each one composed of the following subfields (`id`/`idPattern` or
-  `type` must be present):
-    + `id` or `idPattern`: Id or pattern of the affected entities. Both cannot be used at the same
-      time, but at least one of them must be present.
-    + `type`: Type of the affected entities (optional).
++ `entities`: A list of objects, each one composed of the following subfields:
+  + `id` or `idPattern`: Id or pattern of the affected entities. Both cannot be used at the same
+     time, but at least one of them must be present.
```
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: at least one of them ...
at least one, but maximum one?
Doesn't make much sense:

    Both cannot be used at the same time, but one of them must be present.

or, more concise:

    Mandatory and mutually exclusive

Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Note that the text "Both cannot be used at the same time, but one of them must be present" is already used in the previous version. In the PR, I have just moved it a bit.

However, I understand that your suggestion should be moved to the the fiware-ngsiv2-reference.errata to be included in the "proof read" for the next RC.

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 2]
```diff
@@ -899,12 +899,12 @@ to keep your client decoupled from implementation details.
 
 # Group Entities
 
-### List entities [GET /v2/entities{?limit,offset,options,type,id,idPattern,q,georel,geometry,coords,attrs,orderBy}]
+### List entities [GET /v2/entities{?limit,offset,options,type,id,idPattern,typePattern,q,georel,geometry,coords,attrs,orderBy}]
 
-Retrieves a list of entities that match different criteria (by id, idPattern, type or those which
+Retrieves a list of entities that match different criteria (by id, idPattern, type, typePattern or those which
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: maybe better 

"by pattern matching (over id or type)"

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 3]
```diff
@@ -899,12 +899,12 @@ to keep your client decoupled from implementation details.
...
 match a query or geographical query). A given entity has to match all the criteria to be retrieved
 (i.e., the criteria is combined in a logical AND way). Note that the id and idPattern parameters are
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Probably better to say

"pattern matching query parameters are incompatible with their corresponding exact matching parameters, i.e. idPattern with `id` and `typePattern` with `type`.

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 4]
```diff
@@ -923,9 +923,13 @@ Response code:
     
     + type: Room (optional, string) -  comma-separated list of elements.
       Retrieve entities whose type matches one of the elements in the list.
+      Incompatible with typePattern.
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: nit: use `typePattern`

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 5]
```diff
@@ -923,9 +923,13 @@ Response code:
...
 
     + idPattern: Bode_.* (optional, string) - A correctly formated regular expression.
       Retrieve entities whose ID matches the regular expression. Incompatible with id.
+
+    + typePattern: Room_.* (optional, string) - A correctly formated regular expression.
+      Retrieve entities whose type matches the regular expression. Incompatible with type.
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: nit: use `type`

**Comment by jmcanterafonseca** [comment]
LGTM with the comments and nits addressed

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 1 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 9e76bf5

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 2 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 9e76bf5

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 3 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 9e76bf5

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 4 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 9e76bf5

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 5 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 9e76bf5

**Comment by kzangeli** [comment]
LGTM

**Comment by crbrox** [comment]
LGTM

**Pull request merged**